arangod/Transaction/Methods.cpp

```diff
@@ -1062,15 +1061,55 @@ Future<OperationResult> transaction::Methods::insertLocal(std::string const& cna
     TRI_ASSERT(!isPrimaryKeyConstraintViolation || !key.isNone());
 
     bool didReplace = false;
+
+#ifdef USE_ENTERPRISE
+    if (collection->isSmart() && collection->type() == TRI_COL_TYPE_EDGE &&
+        ServerState::instance()->isSingleServer()) {
+      auto vsecol =
+          dynamic_cast<arangodb::VirtualSmartEdgeCollection const*>(collection.get());
+
+      if (vsecol == nullptr) {
+        // Cast did not work. Illegal state
+        return Result(TRI_ERROR_NO_SMART_COLLECTION);
+      }
+
+      auto vRes = vsecol->verifyEdge(value);
+      if (vRes.fail()) {
+        return vRes;
+      }
+    }
+#endif
+
     if (!isPrimaryKeyConstraintViolation) {
-      // regular insert without overwrite option. the insert itself will check if the
-      // primary key already exists
+      // regular insert without overwrite option. the insert itself will check
+      // if the primary key already exists
+#ifdef USE_ENTERPRISE
+      if (collection->isSmart() && collection->type() == TRI_COL_TYPE_DOCUMENT &&
```

Comment:
Why is this code here?
We either need the same on the Cluster (so we move it to the insert path before we go to InsertLocal) or not at all.

Reply:
Yes this will be part of the SmartVertex Schema Validator refactoring which is currently ongoing.